libraries/AP_KDECAN/AP_KDECAN.cpp

```diff
@@ -53,603 +41,282 @@ const AP_Param::GroupInfo AP_KDECAN::var_info[] = {
     AP_GROUPEND
 };
 
-const uint16_t AP_KDECAN::SET_PWM_MIN_INTERVAL_US;
-
 AP_KDECAN::AP_KDECAN()
 {
     AP_Param::setup_object_defaults(this, var_info);
-
-    debug_can(AP_CANManager::LOG_INFO, "constructed");
-}
-
-AP_KDECAN *AP_KDECAN::get_kdecan(uint8_t driver_index)
-{
-    if (driver_index >= AP::can().get_num_drivers() ||
-        AP::can().get_driver_type(driver_index) != AP_CANManager::Driver_Type_KDECAN) {
-        return nullptr;
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_KDECAN must be singleton");
     }
-    return static_cast<AP_KDECAN*>(AP::can().get_driver(driver_index));
+#endif
+    _singleton = this;
 }
 
-bool AP_KDECAN::add_interface(AP_HAL::CANIface* can_iface) {
-
-    if (_can_iface != nullptr) {
-        debug_can(AP_CANManager::LOG_ERROR, "Multiple Interface not supported");
-        return false;
+void AP_KDECAN::init()
+{
+    if (_driver != nullptr) {
+        // only allow one instance
+        return;
     }
 
-    _can_iface = can_iface;
-
-    if (_can_iface == nullptr) {
-        debug_can(AP_CANManager::LOG_ERROR, "CAN driver not found");
-        return false;
-    }
+    AP_CANManager::Driver_Type protocol = AP_CANManager::Driver_Type_None;
 
-    if (!_can_iface->is_initialized()) {
-        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized");
-        return false;
+    for (uint8_t i = 0; i < HAL_NUM_CAN_IFACES; i++) {
+#if HAL_CANMANAGER_ENABLED
+        protocol = AP::can().get_driver_type(i);
+#elif defined(HAL_BUILD_AP_PERIPH)
+        protocol = CANSensor::get_driver_type(i);
+#endif
+        if (protocol == AP_CANManager::Driver_Type_KDECAN) {
+            _driver = new AP_KDECAN_Driver();
+            return;
+        }
     }
+}
 
-    if (!_can_iface->set_event_handle(&_event_handle)) {
-        debug_can(AP_CANManager::LOG_ERROR, "Cannot add event handle");
-        return false;
+void AP_KDECAN::update()
+{
+    if (_driver == nullptr) {
+        return;
     }
-    return true;
+    _driver->update((uint8_t)_num_poles.get());
 }
 
-void AP_KDECAN::init(uint8_t driver_index, bool enable_filters)
+AP_KDECAN_Driver::AP_KDECAN_Driver() : CANSensor("KDECAN")
 {
-    _driver_index = driver_index;
+    register_driver(AP_CANManager::Driver_Type_KDECAN);
```

Comment:
This can fail, and you're not checking that.  But, then, nobody else is either, so whatever.